homeassistant/helpers/script.py

```diff
@@ -444,13 +444,11 @@ async def async_run(self) -> ScriptRunResult | None:
 
         try:
             self._log("Running %s", self._script.running_description)
-            for self._step, self._action in enumerate(self._script.sequence):
-                if self._stop.done():
-                    script_execution_set("cancelled")
-                    break
-                await self._async_step(log_exceptions=False)
-            else:
-                script_execution_set("finished")
+            async with async_interrupt.interrupt(self._stop, ScriptStoppedError, None):
+                for self._step, self._action in enumerate(self._script.sequence):
+                    await self._async_step(log_exceptions=False)
```

Comment:
One thing that is different in behavior is that we now will stop a step during execution, instead of after. I know services are shielded, so that won't be impacted.

Should we update `_async_futures_with_timeout` to no longer wait for `self._stop` ?

Reply:
Yeah its not needed anymore.  I took it out and clean up a few more places

Reply:
I think we can get rid of run long action as well

Reply:
The delayed step can be further simplified

Reply:
I adjusted everything that has a unsub to make sure it has a finally block and there are no awaited in between when its created and when the finally happens to ensure its unsubed on cancellation.

attaching triggers was not cancellation safe, so I fixed triggers as well since it needs to be